File: solutions/054_permute_unique.go
package main

import "sort"

// 不推荐我的解法 我的解法只是比较直观
// 但很耗内存
// 此题和三数之和避免重复答案方法一样
func permuteUnique(nums []int) [][]int {
	rtv := [][]int{}
	sort.Ints(nums) // 先排序，这样能跳过重复的数字

	// options代表可供填入数字 permute是已填入的数字组合
	var backtrace func([]int, []int)
	backtrace = func(options, permute []int) {
		if len(options) == 0 {
			rtv = append(rtv, permute)
			return
		}
		i := 0
		for i < len(options) {
			newOptions := []int{}
			for j := 0; j < len(options); j++ {
				if j != i {
					newOptions = append(newOptions, options[j])
				}
			}
			newPermute := append([]int{}, permute...)
			newPermute = append(newPermute, options[i])
			backtrace(newOptions, newPermute)
			value := options[i]
			for i < len(options) && options[i] == value {
				i++ // 这里我们要跳过重复的数字 这样就能避免重复答案
			}
		}
	}

	backtrace(nums, []int{})
	return rtv
}

File: solutions/__054_permute_unique.cpp
/*
 *      Author: peanutzhen
 *      Created time: 2021/9/1 21:19:55
 */

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> ans;        // 保存答案
    vector<bool>        visited;    // 该数字是否被填入
    vector<int>         permute;    // 一种排列的可能
    int                 n;          // 数字数量

    void backtrace(vector<int> &nums, int depth) {
        if (depth == n) {
            ans.push_back(permute);
            return;
        }
        for (int i = 0; i < n; ++i) {
            // 去重
            if (!visited[i] && (i == 0 || nums[i] != nums[i-1] || visited[i-1])) {
                visited[i] = true;
                permute[depth] = nums[i];
                backtrace(nums, depth+1);
                visited[i] = false;
            }
        }
    }

    vector<vector<int>> permuteUnique(vector<int>& nums) {
        ans.clear();
        n = (int) nums.size();
        sort(nums.begin(), nums.end());
        visited = vector<bool>(n, false);
        permute = vector<int>(n);
        backtrace(nums, 0);
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    auto nums = vector<int>{1,1,2};
    auto s = new Solution();
    s->permuteUnique(nums);
    return 0;
}

